Source/Core/Core/Config.h

```diff
@@ -0,0 +1,185 @@
+// Copyright 2017 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <string>
+
+#include "Common/Config/Config.h"
+#include "Common/Config/Enums.h"
+
+namespace Config
+{
+struct ConfigLocation
+{
+  System system;
+  std::string section;
+  std::string key;
+
+  bool operator==(const ConfigLocation& other) const;
+  bool operator!=(const ConfigLocation& other) const;
+  bool operator<(const ConfigLocation& other) const;
+};
+
+template <typename T>
+struct ConfigInfo
+{
+  ConfigLocation location;
+  T default_value;
+};
+
+template <typename T>
+T Get(LayerType layer, const ConfigInfo<T>& info)
+{
+  return GetLayer(layer)
+      ->GetOrCreateSection(info.location.system, info.location.section)
+      ->template Get<T>(info.location.key, info.default_value);
+}
+
+template <typename T>
+T Get(const ConfigInfo<T>& info)
+{
+  return Get(LayerType::Meta, info);
+}
+
+template <typename T>
+T GetBase(const ConfigInfo<T>& info)
+{
+  return Get(LayerType::Base, info);
+}
+
+LayerType GetActiveLayerForConfig(const ConfigLocation&);
+
+template <typename T>
+LayerType GetActiveLayerForConfig(const ConfigInfo<T>& info)
+{
+  return GetActiveLayerForConfig(info.location);
+}
+
+template <typename T>
+void Set(LayerType layer, const ConfigInfo<T>& info, const T& value)
+{
+  GetLayer(layer)
+      ->GetOrCreateSection(info.location.system, info.location.section)
+      ->Set(info.location.key, value);
+  InvokeConfigChangedCallbacks();
+}
+
+template <typename T>
+void SetBase(const ConfigInfo<T>& info, const T& value)
+{
+  Set<T>(LayerType::Base, info, value);
+}
+
+template <typename T>
+void SetCurrent(const ConfigInfo<T>& info, const T& value)
+{
+  Set<T>(LayerType::CurrentRun, info, value);
+}
+
+template <typename T>
+void SetBaseOrCurrent(const ConfigInfo<T>& info, const T& value)
+{
+  if (GetActiveLayerForConfig(info) == LayerType::Base)
+    Set<T>(LayerType::Base, info, value);
+  else
+    Set<T>(LayerType::CurrentRun, info, value);
+}
+
+// Configuration Information
+
+// Graphics.Hardware
+
+extern const ConfigInfo<bool> GFX_VSYNC;
```

Comment:
It's not that big of a deal but I'm guessing all the settings will eventually end up here?  I wonder if it makes sense to create a folder structure like:

ConfigurationSettings/
  GraphicsSettings.h
  ControllerSettings.h
   etc?